lib/srv/desktop/rdp/rdpclient/src/scard.rs

```diff
@@ -177,7 +179,13 @@ impl Client {
             .contexts
             .get(req.common.context.value)
             .ok_or_else(|| invalid_data_error("unknown context ID"))?;
-        let handle = ctx.connect(req.common.context, self.uuid, &self.cert_der, &self.key_der)?;
+        let handle = ctx.connect(
+            req.common.context,
+            self.uuid,
+            &self.cert_der,
+            &self.key_der,
+            self.pin.clone(),
```

Comment:
I don't think it's worthwhile to eliminate this `.clone()` but as a mild stye nitpick, I usually prefer to keep owned types to structs and (only where necessary) constructors. In other words, pass the pin as a `&str` to `ctx.connect()` and call `.clone()` on that as late as possible (ideally as the struct is being built).

This helps make sure we don't create unnecessary clones elsewhere as things get passed around.

Reply:
The Rust API guidelines disagree with this: https://rust-lang.github.io/api-guidelines/flexibility.html#caller-decides-where-to-copy-and-place-data-c-caller-control
> If a function requires ownership of an argument, it should take ownership of the argument rather than borrowing and cloning the argument.

Reply:
I think Tim is pointing out that connect here doesn't actually require ownership of the pin because it just passes it through to the piv::Card.